[1주차 과제] 정유빈

youbeen2798 opened the pull request:
## 이번 주에 어떤 작업을 했는지 설명해주세요.


<br>

## 특히 어떤 부분을 리뷰받고 싶나요?


<br>

## 이번 주는 어떻게 학습했나요? 아래 질문에 짧게 답변주세요!

### 이번 주에 학습에 투자한 시간

 - 

### 학습 하면서 좋았던 점과 아쉬웠던 점

 - 
 - 

### 어려움을 겪는 부분

 -

### 스터디 개선되었으면 하는 점

 -

youbeen2798 committed:
Create [1주차 과제] 정유빈.md

youbeen2798 committed:
Update [1주차 과제] 정유빈.md

YOONSEOUNGHWAN commented:
##FeedBacek_윤승환

---

1. **Service 레이어는 인터페이스를 따로 두지 않는 경우가 많다고 알고 있는데, 그 이유가 궁금합니다.**
Serivce 계층은 데이터를 가공하여 원하는 비즈니스 로직을 만족하는 곳입니다.
이상적으로는 모든 계층은 Interface를 부여하여 개발하는 것이 좋습니다. OCP(Open Close Principle)참고.
인터페이스를 만든다는 것을 **추상화**한다고 생각하면 좋을 것 같아요!<br>
**하지만** 교체하거나, 확장가능성이 없는 클래스에 Interface를 적용한다면 그에 따른 추상화 비용이 발생하기에, 초기 빠른 개발에서는 우선 구체클래스를 구현하고 -> 리팩토링 과정에서 추상화해도 괜찮습니다.
->저는 개인적으로, OOP(객체지향프로그래밍)을 지향하기 위해서 **추상화비용**이 많이 들어도 괜찮다고 생각합니다..
*추상화 비용 : 바로 로직을 생각하지 않고, 해당 객체가 어떤 메소드(행동)를 제공해야할 지에 대해서 드는 비용.. 아닐수도있음.*

2.   **로그인하는 사용자가 여러 명일 때는 어떤 식으로 동작하는지 잘 모르겠습니다.**
DB와 연결하기 위해서
.setinitialSize로 DB와 접속할 객체를 생성합니다.
.setMaxTotal로 DB와 접속할 최대 객체를 제한합니다.
만약 질문주신 것처러 10개의 객체를 생성했는데, 동시 요청이 15개가 온다면 15개의 DB연결 객체가 생성될 것 입니다.
만약 10개의 객체를 생성했는데, 동시 요청이 30개가 온다면 10개의 요청은 커넥션이 반납될때까지 기다릴 것 입니다.

- 커넥션을 생성하는 것은 굉장히 느린 작업입니다. 따라서 앞선 설정처럼 미리 ConnectionPool에 Connection을 등록해두면 불필요한 작업을 생략할 수 있습니다.

- 또한 하나의 Connection은 하나의 트랜잭션을 관리한다고 생각하면 되고,
요청이 끝난 Connection은 ConnectionPool에 반납됩니다. 이로 인하여 Connection을 재사용하므로써 생성 비용을 절감합니다.

- 추가로, ConnectionPool의 모든 Connection이 요청을 받았고 여유 Connection이 없는 상태에서 요청이 왔다면
대기장소(Queue)에서 Polling하고 있다가 Connection이 반납되기를 기다립니다.

SeungYongChoi reviewed (approved):
저도 조금 도움을 드리자면
connection이란 DB와 통신하기 위해 **연결**을 맺는 것을 의미합니다.
basicDataSource.setInitialSize(10); -> 기본적으로 10개의 연결을 가지고 있겠다.
즉 미리 DB와의 연결을 10개 가지고 있겠다.
basicDataSource.setMaxTotal(20); -> 최대 20개까지 연결을 하겠다.
즉 DB와의 연결 갯수를 최대 20개로 제한하겠다.
라고 생각합니다.

말씀 주신 로그인하는 사용자로 예시를 들어 보겠습니다.
로그인 하기 위해서는 
1. ID, Password를 입력받는다.
2. DB에 저장된 데이터와 비교한다.
3. 일치 여부에 따라 특정 값을 return한다.
이렇게 3가지로 생각해 보겠습니다.

2번 과정 중 DB에 저장된 아이디와 비밀번호를 가져오는 과정에서 DB와의 연결이 필요하게 됩니다.
이때 connection을 하나 가져오게 됩니다.
`설정에 따라 connection을 놓는 시점이 달라지긴 하는데 이 부분 궁금하시면 말씀해 주세요 다시 답글 남겨드리겠습니다.`

- 만약 connection을 가져오는 과정에서 이미 10개의 connection이 사용되고 있다면 DataSource는 새로운 connection을 만든 후(새로운 connection을 생성하는 것은 리소스가 큰 즉 굉장히 느린 작업입니다.) 할당해 줍니다.(최대 20개까지)
- 만약 connection을 가져오는 과정에서 이미 20개의 connection이 사용되고 있다면 다른 누군가가 사용중인 connection을 놓기 전까지 대기장소(Queue)에서 기다리고 있다가 놓은 후 connection을 할당받게 됩니다.

youbeen2798 committed:
Add files via upload

디비 연결 설정인 UnsatisfiedDependencyException 에러가 나는데, 해결하지 못하여서 제대로 진도가 나가지 못하였습니다ㅠㅠ 
 Error creating bean with name 'dataSourceScriptDatabaseInitializer' defined in class path resource 에러가 나는데 혹시 어떻게 해결해야 할지 모르겠습니다.

youbeen2798 committed:
Create [3주차 과제] 정유빈

youbeen2798 committed:
Rename [3주차 과제] 정유빈 to [3주차 과제] 정유빈..md

youbeen2798 committed:
Rename [3주차 과제] 정유빈..md to [3주차 과제] 정유빈.md

youbeen2798 committed:
Merge pull request #1 from DKU-D-Coding/main

Create CODEOWNERS

youbeen2798 committed:
Create [3주차 과제]    정유빈.md

youbeen2798 committed:
Merge pull request #2 from youbeen2798/youbeen2798-patch-1

Create [3주차 과제]    정유빈.md

youbeen2798 committed:
.

..

youbeen2798 reviewed (commented):
.

youbeen2798 committed:
Revert "Create [3주차 과제]    정유빈.md"

youbeen2798 committed:
Merge pull request #3 from youbeen2798/revert-2-youbeen2798-patch-1

Revert "Create [3주차 과제]    정유빈.md"

youbeen2798 committed:
Update [3주 차 과제] 정유빈.md

youbeen2798 committed:
Merge pull request #4 from youbeen2798/youbeen2798-patch-1

Update [3주 차 과제] 정유빈.md

won983212 commented:
과제 pr올리실 때 아래와 같이 해주세요. 
1. 브랜치를 하나 새로 판다. (예시: 3주차 과제라는 의미로 youbeen2798-3) 
2. 새로 생성한 브랜치에서 이 레포지토리로 pr을 올린다.

이렇게하시면 pr이 올라올거에요!
